승자확인

GangWoon opened the pull request:
알고리즘 너무 어려워요.... 하루종일 작성했습니다.. ㅠㅜ

GangWoon committed:
의존성 주입

뷰 객체에서 모델 객체를 모르게 했습니다.

GangWoon committed:
컨밴션 수정

GangWoon committed:
카드와 덱 파일 생성

햇갈려서 분리했습니다.

GangWoon committed:
Card 구조체 변경

hands에서 card를 사용해야하기 때문에 변경했습니다.

GangWoon committed:
Hands 구조체 수정

텍사스 룰을 반영했습니다.

GangWoon committed:
Hands 구조체 수정

player가 사용할 수 있게 수정했습니다.

GangWoon committed:
Player 클래스 수정

players에서 사용할 수 있도록 수정했습니다.

GangWoon committed:
Players 구조체 수정

player들을 비교할 수 있게 수정했습니다.

GangWoon committed:
이미지 추가

GangWoon committed:
GameTableStackView 수정

승리표시하는 이미지 뷰 하나 추가했습니다.

GangWoon committed:
Game 클래스 수정

적용을 안해서 메소드를 사용해서 적용시켰습니다.

godrm commented:
커밋 로그나 PR 메시지도 좀 더 상세하게 작성해보세요. 
별거 아닌것 같아도 작성한 코드와 함께 기록으로 남기는 겁니다. 
그만큼 다른 개발자들과 소통하는 방법중에 하나구요. 
요즘은 인터뷰할 때 커밋 로그도 보는 회사가 많습니다. 
개발 과정에서의 생각과 의사결정 과정을 표현해보면 좋겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -19,69 +18,34 @@ class Card {
+            let isEmpty = dictionary[self.rawValue] == nil
+            let updateValue = (dictionary[self.rawValue] ?? 0) + 1
+            dictionary.updateValue(isEmpty ? 1 : updateValue , forKey: self.rawValue)
+        }
```
여기 dictionary가 Rank에 왜 필요한지 모르겠어요. 
Rank가 같은 카드의 개수를 표현하려고 한 것 같은데, 상위 모델에서 처리할 개념을 여기에 추가한 것 같네요. 
그리고 Dictionary를 사용하려면 Rank를 그대로 타입으로 활용하세요.

GangWoon replied:
제가 하위 모델에서 처리를 했던 이유가 rawValue를 직접적을 사용 가능해서 작성했었는데 좋은 방법은 아니였군요!! 참고하겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/Hands.swift:
```diff
@@ -10,15 +10,59 @@ import Foundation
+                self = .fourCard
+            default:
+                self = .none
+            }
```
enum 값만 조정하면 여기 switch-case 안쓸 수 있을것 같은데요?

godrm commented on CardGameApp/CardGameApp/Model/Hands.swift:
```diff
@@ -10,15 +10,59 @@ import Foundation
+        }
+        
+        func findPlayer(with value: Int) -> Bool {
+            value == self.rawValue
```
점수를 계산하는 곳에서 findPlayer는 맥락이 맞지 않아서 어색한데요? 여기 player가 있던가요?

godrm commented on CardGameApp/CardGameApp/Model/Hands.swift:
```diff
@@ -10,15 +10,59 @@ import Foundation
+        }
+        
+        func compare(with value: Int) -> Int {
+            value < self.rawValue ? self.rawValue : value
```
여기는 뭐와 뭐를 비교하는거죠? 
TexasHoldemRule와  비교하는 거면 Int 대신 TexasHoldemRule를 써야 할 것 같아요

godrm commented on CardGameApp/CardGameApp/Model/Hands.swift:
```diff
@@ -27,8 +71,42 @@ struct Hands {
+        var grade: TexasHoldemRule = .none
+        grade = applyRules(dictionary: dictionary)
+        findTwoPair(grade: &grade, dictionary: dictionary)
+        findStraight(grade: &grade, dictionary: dictionary)
```
grade와 dictionary 변수는 여러 메소드들에서 여러번 쓰는 데이터라면 inout 보다 인스턴스 변수(프로퍼티)로 만드는 게 어떨까요?
grade를 input 과 output으로 구분해보는 것도 가능하지 않을까요?

godrm commented on CardGameApp/CardGameApp/Model/Hands.swift:
```diff
@@ -27,8 +71,42 @@ struct Hands {
+        var count = 0
+        let twoPair = 2
+        dictionary.values.forEach { $0 == twoPair ? count += 1 : nil }
+        grade == .pair && count > 1 ? grade = .twoPair : nil
```
위에도 얘기했지만 이전 grade에 따라서 상태가 달라지는 거면, inout 보다는 return 값을 활용하는 게 어떨까요? 
private 이라 테스트는 안해도 되겠지만, 함수를 순수함수로 만드는 것은 매우 중요합니다. 참고하세요.

GangWoon replied:
네 순수함수를 만드려고 노력하겠습니다!

godrm commented on CardGameApp/CardGameApp/Model/Player.swift:
```diff
@@ -12,12 +12,19 @@ class Player {
     
     private var cards: Hands = Hands()
     private var identifier: String
+    var winner: Bool = false
```
winner 값은 외부에서 바꿀 수 있는 값인가요?

godrm commented on CardGameApp/CardGameApp/Model/Player.swift:
```diff
@@ -38,4 +45,16 @@ class Player {
     }
+    
+    func checkWinner() {
+        winner.toggle()
```
checkWinner는 winner를 찾기만 하는건지, winner값을 바꾸는 건지, winner를 가져오는건지 명확하지 않습니다.
check-보다는 더 구체적으로 명확한 의도를 표현해보세요.

godrm commented on CardGameApp/CardGameApp/Model/Player.swift:
```diff
@@ -38,4 +45,16 @@ class Player {
+    }
+    
+    func findPlayer(with value: Int) -> Bool {
+        cards.findPlayer(with: value)
```
player.findPlayer(with:) 는 참가자한테 다른 참가자를 찾는건가요? 이 함수의 의도를 모르겠습니다.

GangWoon replied:
메소드 이름 작성하기 전에 조금 더 노력하겠습니다 ㅋㅋ

godrm commented on CardGameApp/CardGameApp/Model/Players.swift:
```diff
@@ -29,4 +29,38 @@ struct Players {
+            players.sorted {
+                $0.winner = false
+                $1.winner = false
+                return $0.topValue > $1.topValue
```
정렬하는 로직에서 값을 바꾸기까지 하는건 부작용을 일으키기 좋은 방식입니다. 지양하세요.

godrm commented on CardGameApp/CardGameApp/Model/Players.swift:
```diff
@@ -29,4 +29,38 @@ struct Players {
+                return $0.topValue > $1.topValue
+            }.first?
+                .checkWinner()
+        }
```
여기있는 비교하는 로직이야 말고 player 내부로 옮길 수 있을 것 같네요. topValue를 가져오지 않고 `$0 > $1` 이렇게 비교하도록 개선해보세요.

GangWoon replied:
비교해야 하는 대상이 TexasHoldemRule, topValue라서 TexasHoldemRule을 > 비교 연사자를 통해서 처리했고 topValue는 여전히 .topValue로 처리했습니다~~

godrm commented on CardGameApp/CardGameApp/View/GameTableStackView.swift:
```diff
@@ -10,14 +10,20 @@ import UIKit
-    private var game: Playable? = Game.fiveCardStud(gamers: Players())
-    private var stackView: UIStackView!
+    private var game: Playable?
+    private var stackView: UIStackView?
```
GameTableStackView가 이미 UIStackView를 상속받았는 데, 내부에 stackView가 더 필요한건가요?
자기 자신에 추가하면 되는거 아니었나요? 다시 생각해보세요.

GangWoon replied:
아직도 놓치는 부분이 많은 거 같아요!!

godrm reviewed (commented):
객체 역할 분담에 대해 피드백을 확인해보세요.
상위 개념은 상위 모듈에, 하위 개념은 하위 모듈에 담아서 표현하려고 해보세요.
inout 보다는 매개변수와 return 값을 활용해보세요.

GangWoon committed:
GameTableStackView 수정

클래스 자체에서 UIStackView을 상속받고 있는데 UIView를 사용하는 것처럼 UIStackView를 생성해서 저장하고 있던 부분을 클래스 자체에 저장하는 방식으로 변경했습니다.

GangWoon committed:
Card 구조체 수정

카드에서 딕셔너리를 만드는게 아닌 스트레이트를 계산할때 연속된 숫자인지만 확인할 수 있게 변경했습니다. 그리고 player에서 rank를 사용해야되서 value라는 컴퓨티트 프로퍼티를 만들었습니다. 이렇게 처리하기 싫어서 다른 방법을 생각해봤었습니다. description에서 처음 문자 지우고 J Q K 는 분기로 처리하려 했지만 너무 더러워지는 느낌을 받아서 어쩔 수 없이 사용했습니다!

GangWoon committed:
Hands 구조체 수정

처음 생각할 때 hands에 정보를 갖고 있으면 처리하기 편해서 그렇게 했었는데 구조에 대해서 조금 더 생각해보려고 노력하겠습니다. topValue는 TexasHoldemRule이 같을때 처리하기 위해서 만들었습니다. 여기서 고민했던 점은 topValue가 hands에서 갖고 있는게 아닌 player가 hands.sorted { $0 > $1 }.first?.value로 사용하고 싶었는데 sorted() 메소드가 어떻게 구현할 지 몰라서 포기했습니다. 혹시 ios framework의 코드를 볼 수 있는 방법이 있을까요??  jump to definition에 가도 자세히 안나와 있어서 궁금합니다!!

GangWoon committed:
Player 클래스 수정

TexasHoldemRule을 player로 옮기고 그에 계산하는 방법을 추가했습니다.
comparable 프로토콜을 채택해서 player들끼리 TexasHoldemRule를 바로 비교할 수 있게 만들었습니다.
inout을 써서 처리했던 부분을 return하는 형태로 바꿔서 처리했습니다. winner 속성을 private로 바꿔서 처리했습니다.

GangWoon committed:
GameTableStakcView 수정

winner 속성이 private로 변하면서 코드를 수정했습니다.

GangWoon committed:
players 클래스 수정

메소드를 줄이고 조금 더 명시적인 이름을 사용했습니다.

GangWoon committed:
Deck 구조체 수정

강제 언래핑을 사용하는 부분을 수정했습니다.

godrm commented on CardGameApp/CardGameApp/Model/Players.swift:
```diff
@@ -16,6 +16,20 @@ struct Players {
+    private func choiceTopValue() {
+        players.forEach { $0.resetWinnerState() }
+        let topValue = players.sorted { $0.topValue > $1.topValue }.first ?? Player(identifier: "")
+        topValue.toggleWinner()
```
제 의견은 그렇습니다. 
player 중에 누가 승자인지 처리하는 로직을 player에 winner 값을 갖고 있기 보다는 여기서 player 를 비교만 하고, 
누가 승자인지는 Players 내부에 (이름도 topValue보다는) topPlayer를 속성으로 갖고 있으면 되지 않을까 싶습니다. 그러면 내부에 winner 속성도 필요없고, toggleWinner() 같은 메소드도 필요없죠. 
Player를 비교할 때는 내부 topValue에 접근하지말고 Player끼리 비교하는 게 좋습니다. 
참고하세요.

godrm commented on CardGameApp/CardGameApp/View/GameTableStackView.swift:
```diff
@@ -42,24 +41,22 @@ class GameTableStackView: UIStackView {
+        subviews.forEach {
+         removeArrangedSubview($0)
             $0.removeFromSuperview()
         }
```
들여쓰기가 달라서 다른 건줄 알았네요. 들여쓰기도 잘 맞춰주세요.

godrm commented:
단지 동작하는 코드만으로 다른 개발자를 설득하기 어렵습니다. 
세부 사항에서 꼼꼼하게 왜 그렇게 설계했는지 설명할 수 있어야 합니다. 
피드백 내용 중에 명확하지 않은 게 있으면 다시 댓글을 남겨서 이해하고 넘어가세요.
수고하셨습니다.
